### ISSUE TITLE
how to control record speed

### ISSUE THREAD

**Issue by ggslayer** [bytedeco/javacv#1916]
Hi, I'm working on rtmp push program with javacv recently, now, simply create a demo read local file to rtmp, code like below, It works ok.  Because I don't have many a/v process knowledge before, I'm worried about some detail(which I don't know about) when the project upto prod env in the future. Please help me about My worries: 

1. If the framerate is 30/s, how to control the push speed to rtmp server, now I simply make a sleep, I think it's not a correct method.
    I want to control speed, because, I don't want to push faster than 30/s( this will cause some problem on rtmp server or player terminal? ), I also don't want to push slower than 30/s( I think this will cause block on player), Is my opinion is correct? If yes, How to control this record speed?  If not, please tell me the recommened practice.

2. I notice that  the grabber 、recorder 、frame all have timestamp can be get/set,  but no time_base/pts fields in ffmepg c api, So, I think I control timestamp only to do all my time relation work is ok?  for example: audio/video sync, push speed, ...

3. In my scenc, I need to play multi video files looply 7*24, If every recorded frame's timestamp should be set correctly, after days, the frame's timestamp should be set to a very large num, Is my opinion correct?

Thank  you for your  patience, waiting for your reply.

`

    FFmpegLogCallback.set();

    frameGrabber = new FFmpegFrameGrabber(srcfile);
    frameGrabber.setPixelFormat(avutil.AV_PIX_FMT_RGBA);

    Frame captured_frame = null;

    try {
        frameGrabber.start();

        recorder = new FFmpegFrameRecorder(destfile, frameGrabber.getImageWidth(), frameGrabber.getImageHeight(), frameGrabber.getAudioChannels());
        recorder.setInterleaved(true);
        recorder.setVideoOption("tune","zerolatency");
        recorder.setVideoOption("preset", "ultrafast");
        recorder.setVideoCodec(avcodec.AV_CODEC_ID_H264);
        recorder.setFormat("flv");
        //recorder.setPixelFormat(avutil.AV_PIX_FMT_RGBA);   // this is not supported.
        recorder.setFrameRate(frameGrabber.getFrameRate());
        recorder.setVideoBitrate(frameGrabber.getVideoBitrate());

        recorder.setAudioBitrate(192000);
        recorder.setAudioOptions(frameGrabber.getAudioOptions());
        recorder.setAudioQuality(0);
        recorder.setSampleRate(44100);
        recorder.setAudioCodec(avcodec.AV_CODEC_ID_AAC);
        recorder.start();

        frameGrabber.setTimestamp(0);

        while (true) {
            try {
                captured_frame = frameGrabber.grabFrame();

                if (captured_frame == null) {
                    System.out.println("!!! Failed cvQueryFrame");
                    break;
                }

                if(captured_frame.type == Frame.Type.VIDEO){
                    captured_frame = MixFrameDraw(captured_frame); 
                    Thread.sleep(20);                                                             // 2
                }else{
                    //System.out.println("audio frame");
                }

                recorder.record(captured_frame, AV_PIX_FMT_ARGB);          
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public static Frame MixFrameDraw(Frame frame){
        BufferedImage imgFrame = frameConverter.getBufferedImage(frame);

        BufferedImage res = new BufferedImage(bufferImgBg.getWidth(), bufferImgBg.getHeight(), BufferedImage.TYPE_INT_ARGB);
        Graphics graphics = res.getGraphics();
        graphics.drawImage(bufferImgBg, 0, 0, null);
        graphics.drawImage(imgFrame, 100, 100, null);
        graphics.drawImage(bufferImgFg, 100, 400, null);
        graphics.drawImage(bufferImgSku, 100, 1000, null);

        return frameConverter.getFrame(res);
    }
`

**Comment by saudet** [comment]
> 1. If the framerate is 30/s, how to control the push speed to rtmp server, now I simply make a sleep, I think it's not a correct method.
>   I want to control speed, because, I don't want to push faster than 30/s( this will cause some problem on rtmp server or player terminal? ), I also don't want to push slower than 30/s( I think this will cause block on player), Is my opinion is correct? If yes, How to control this record speed?  If not, please tell me the recommened practice.

Yes, that's how it needs to be done. You can try grabAtFrameRate() as well if that does what you need:
https://github.com/bytedeco/javacv/blob/master/src/main/java/org/bytedeco/javacv/FrameGrabber.java#L785

> 2. I notice that  the grabber 、recorder 、frame all have timestamp can be get/set,  but no time_base/pts fields in ffmepg c api, So, I think I control timestamp only to do all my time relation work is ok?  for example: audio/video sync, push speed, ...

There are a lot of pts, dts, time_base files in the C API of FFmpeg. What are you looking for exactly?

>   If the timestamp field is stable? begin with 0 in all audio/video file?

What do you mean "stable"? It doesn't need to start at 0, no.

> 3. In my scenc, I need to play multi video files looply 7*24, If every recorded frame's timestamp should be set correctly, after days, the frame's timestamp should be set to a very large num, Is my opinion correct?

You'll need to check what your server needs. Maybe there's a way to reset it.

> 4. In my code, before frame record, I did't call frame's setTimestamp and the video's frame is 0(return from MixFrameDraw func), In this situation, the program works ok, why?  Is the recorder set the timestamp auto internal? or the timestamp is not importent?

FFmpegFrameRecorder currently ignores Frame.timestamp, if that's what you mean, yes. We can change this, but it won't do anything different than setTimestamp().

**Comment by ggslayer** [comment]
thanks!!   grabAtFrameRate help me,   but it has a problem, the internal startTime field is not reset to 0 when I call grabber's restart method, I copy the grabAtFrameRate code in my test code, control the myself's startTime field, all goes ok when I loop single file.  Thank you!
